src/Microsoft.AspNet.Razor/Generator/Compiler/CodeBuilder/CSharp/CSharpTagHelperCodeRenderer.cs

```diff
@@ -432,38 +359,23 @@ private void RenderEndTagHelpersScope()
                                                   _tagHelperContext.ScopeManagerEndMethodName);
         }
 
-        private void RenderTagOutput(string tagOutputMethodName, bool methodReturnsString)
-        {
-            // If its a string WriteLiteral must be generated.
-            if (methodReturnsString)
-            {
-                CSharpCodeVisitor.RenderPreWriteStart(_writer, _context);
-            }
-            else
-            {
-                RenderStartWriteMethod();
-            }
-
-            _writer.Write(ExecutionContextVariableName)
-                   .Write(".")
-                   .Write(_tagHelperContext.ExecutionContextOutputPropertyName)
-                   .Write(".")
-                   .WriteMethodInvocation(tagOutputMethodName, endLine: false)
-                   .WriteEndMethodInvocation();
-        }
-        private void RenderStartWriteMethod()
+        private void RenderWriteTagHelperMethodCall()
         {
             if (!string.IsNullOrEmpty(_context.TargetWriterName))
             {
                 _writer
-                    .WriteStartMethodInvocation(_context.Host.GeneratedClassContext.WriteToMethodName)
+                    .WriteStartMethodInvocation(_tagHelperContext.WriteTagHelperToAsyncMethodName)
                     .Write(_context.TargetWriterName)
                     .WriteParameterSeparator();
             }
             else
             {
-                _writer.WriteStartMethodInvocation(_context.Host.GeneratedClassContext.WriteMethodName);
+                _writer.WriteStartMethodInvocation(_tagHelperContext.WriteTagHelperAsyncMethodName);
             }
+
+            _writer.Write(ExecutionContextVariableName)
+                   .WriteEndMethodInvocation(endLine: false)
+                   .WriteLine(".Wait();");
```

Comment:
teensy nit: here's where my commit comment lands.  you're indenting the easy and easy-to-maintain way just above (line 366) but doing it the hard way here.